WatIDL - strawman text format for wasm interfaces

wanderer opened the pull request:
## WatIDL
 `/ˈwädl/` :duck: 

WatIDL is strawman text format for Wasm interfaces. This is an attempt to aggregate feedback from PR https://github.com/WebAssembly/WASI/pull/58

A quick overview
- WatIDL has the object model of capn-proto
- WatIDL has syntax of WAT
- WatIDL needs webIDL bindings

I didn't full flush out how the binding would look like but I think it will be fairly straight since the is a pretty straight forward mapping between watIDL type and webIDL types.

bjorn3 commented on watidl/watidl.md:
```diff
@@ -0,0 +1,118 @@
+While WatIDL follow close to WAT there are some differences
+
+## Types
+WatIDL uses the following primal types
```
s/primal/primitive?

wanderer replied:
thanks, fixed!

syrusakbary commented:
I love this idea!

We've been working in something similar (but much more down to the already-existing types of WebAssembly): [wasm-interfaces](https://github.com/wasmerio/wapm-cli/tree/master/lib/wasm-interface) (*See interfaces proposal in [WebAssembly Design repo](https://github.com/WebAssembly/design/issues/1289)*)

I would love to hear more thoughts about the Wat proposal!

sbc100 commented on watidl/watidl.md:
````diff
@@ -0,0 +1,119 @@
+```
+(func (export "multi-vale")
+  (result $result1 u64)
+  (result $result2 f32)
````
Would it make sense to use the same syntax as the multi-val proposal here: https://github.com/WebAssembly/multi-value/blob/master/proposals/multi-value/Overview.md#functions-with-multiple-return-values

wanderer replied:
i thought this was valid syntax for the multi-val proposal. Well at least [wabt](https://webassembly.github.io/wabt/demo/wat2wasm/) excepts it as valid :stuck_out_tongue: That said if we can find the authoritative definition some where, lets follow that!

devsnek replied:
`(param i32) (param i32) (result i32) (result i32)`, `(param i32 i32) (result i32 i32)`, and permutations of both, are all valid.

sbc100 commented on watidl/watidl.md:
```diff
@@ -0,0 +1,119 @@
+    (result u64)
+  )
+
+  (func (export "not-static")
```
Maybe leave func with is current meaning and use "method" when you want an implicit "this" for 0?

wanderer replied:
yep  i see how that part is confusing. Method sounds like an accurate description.

sbc100 commented on watidl/wasi-core.watidl:
```diff
@@ -0,0 +1,1076 @@
+
+	;; Move the offset of a file descriptor.
+	;; Note: This is similar to lseek in POSIX.
+	(func (export "seek")
```
So is "fd" a separate WASI interface or just part of wasi-core?    Right now we just have wasi-unstable as a single interface.     If we wanted to literally translate the current wasi-unstable as it stands would be just one interface?

Maybe this isn't a precise representation of wasi-unstable as it is today, but if one wanted to do that presumably one wouldn't try to mode file functions as methods on the 'fd' object like this?  This seems like add OO-style when there was none previously.

wanderer replied:
> So is "fd" a separate WASI interface or just part of wasi-core?

Currently its part of WASI core, but I think there some plans to break up wasi-core into several modules. So this is trying to reflect that future. I was thinking of adding `(interface $fd (export "wasi-unstable")` to explicitly show that all of the imports are under the "wasi-unstable" module namespace.  In the future it would be nice to have multiple modules we could use so I imagine common types would be imported from "wasi-core" or just "wasi" base file descriptor methods would be imported from "wasi-fd" ect. 

> Maybe this isn't a precise representation of wasi-unstable as it is today

I think it would be nice if this could represent wasi-unstable today. I think we can do that by adding some notation to that says "all these interfaces have the same module namespace". That way we can introduce this notation without having to make other changes.

>but if one wanted to do that presumably one wouldn't try to mode file functions as methods on the 'fd' object like this?

So there is another idea here. Yes its very OO-style centric and ideally we can having binding that allow wasm programs to have two different modes of interacting with the interface 1) is the current ADT pattern or 2) have some sort of dynamic dispatching which would allow for virtualization. 

The OO-style would suggest number # 2 but can also express # 1 But if don't have some OO-style notation then how would represent # 2 in the future?

sunfishcode commented on watidl/watidl.md:
````diff
@@ -0,0 +1,119 @@
+WatIDL uses the following primitive types
+
+```
+intType := "i8" | "i6" | "i32" | "i64" | "u8" | "u16" | "u32" | "u64";
````
In underlying wasm, `i32` and `i64` refer to types which don't have a signedness interpretation. Using them to indicate signed types here could lead to ambiguity in some contexts. I suggest using `s32` and similar names for explicitly signed types.

wanderer replied:
updated! thanks

sunfishcode commented on watidl/watidl.md:
````diff
@@ -0,0 +1,119 @@
+```
+intType := "i8" | "i6" | "i32" | "i64" | "u8" | "u16" | "u32" | "u64";
+floatType ::= "f32" | "f64" 
+primType ::= "null" | "bool" | "string" | "data" | intTypes | floatType;
````
Typo: `intTypes` -> `intType`?

wanderer replied:
fixed

sunfishcode commented on watidl/watidl.md:
````diff
@@ -0,0 +1,119 @@
+)
+```
+### static
+A static function is one that does has no associated context. All non static function must bind to a context 
````
Wordo: "does has no" -> "has no".

wanderer replied:
fixed

sunfishcode commented on watidl/watidl.md:
```diff
@@ -0,0 +1,119 @@
+
+(interface $a
+  (func (export "factory")
+    (result $the_result $b)
```
Could you describe what this means in more detail? Would each instance of an interface have its own copy of the exported global variables declared in an interface?

wanderer replied:
> Would each instance of an interface have its own copy of the exported global variables declared in an interface?

thats a good question, and it show that might be some inconsistency here. i'll think about this more

sunfishcode commented on watidl/watidl.md:
```diff
@@ -0,0 +1,119 @@
+    (param u64 u64)
+    (result u64)
+  )
+)
```
Do you intend this to interact with interfaces being first-class? Specifically, are you intending to allow for dynamic polymorphism, where a base class can have multiple derived classes which override methods in the base class, and method dispatch is determined by the dynamic type of an object?

wanderer replied:
so i think i want to back away form interfaces being first-class. I think things will be more simple of interfaces just represent what a module can import and export. I'll try to itterate on this in the next couple of days. But instead of having the OO-style stuff squished on top of an interface (this was capn-proto style) we can put them in to stucts ie

```
(interface $wasi
  ;; all the base methods for fd's such as
  (struct $file_descriptor
    (field $close (func))
    (field $datasync (func))
  )
  ...
)
```

sunfishcode commented on watidl/wasi-core.watidl:
```diff
@@ -0,0 +1,1076 @@
+(type $size_t u64)
```
This is interesting :-). I assume the motivation for using 64-bit here is to be generic between 32-bit and 64-bit. I think that's reasonable, especially since most uses of `size_t` in `wasi_unstable` are replaced here by bindings types like string and array. However, it'd be good to add a comment about this here.

MarkMcCaskey commented:
Following up on today's meeting:

Here's the issue we have for "Wasm Interfaces" which are WAT syntax with only Wasm core types.
https://github.com/WebAssembly/design/issues/1289

We also have a working set of semantics for validating and merging them, but we are of course open to change and compromise for standardization!

@wanderer I have some questions about the `static` vs `dynamic` distinction you made today: if we only use core Wasm types, do we still need this distinction?  I think I didn't fully follow your explanation of why OO-style needs that but ADT-style does not.

sbc100 commented:
I vote for punting on trying to model any kind of dynamic dispatch in the WASI interfaces until we have a real requirement for it (and primitives that might support it).   Keeping this simple and static at the WASI level (at least for now) makes sense to me.

wanderer commented:
@MarkMcCaskey I updated the syntax and removed `static` but added `method` param which was borrow from webidl-binding. I'll read over the dicussion you linked in detail but I think this is very much in the same vain. But we also need away to describe the various data structures that WASI has. Currently this proposal is just using GC like syntax to do that and to rename simple.

> I vote for punting on trying to model any kind of dynamic dispatch in the WASI interfaces until we have a real requirement for it (and primitives that might support it)

@sbc100 what do you think about `method`? I don't mind removing it if ppl think it is unnecessary, but it does give us a clear path forward for once we have dynamic dispatch and (i think) doesn't add to much extra complexity.

sbc100 commented:
> @sbc100 what do you think about `method`? I don't mind removing it if ppl think it is unnecessary, but it does give us a clear path forward for once we have dynamic dispatch and (i think) doesn't add to much extra complexity.

I don't feel super strongly about it, it just seems that its not needed for the current wasi_unstable.  I don't see any problem with modeling `fdstat_get` as function that take a descriptor (on in the future a ref) as arg0.   I'm not sure what we gain by calling it a method and eliding arg0 in the IDL?   Since it will be function of N+1 args at the wasm binary level one could argue it makes IDL<->wasm correspondence less clear.

programmerjake commented:
> > @sbc100 what do you think about `method`? I don't mind removing it if ppl think it is unnecessary, but it does give us a clear path forward for once we have dynamic dispatch and (i think) doesn't add to much extra complexity.
> 
> I don't feel super strongly about it, it just seems that its not needed for the current wasi_unstable. I don't see any problem with modeling `fdstat_get` as function that take a descriptor (on in the future a ref) as arg0. I'm not sure what we gain by calling it a method and eliding arg0 in the IDL? Since it will be function of N+1 args at the wasm binary level one could argue it makes IDL<->wasm correspondence less clear.

one option could be to explicitly have a `self` argument, like Rust or Python.

MarkMcCaskey commented:
> But we also need away to describe the various data structures that WASI has

Hmm, that's a good point -- having data types definitely makes it more useful for things like generating code.

The reason we decided to make something like our current proposal of "Wasm Interfaces" is that we needed a way to allow arbitrary, user-defined ABIs to be combined and modules validated against them to improve the usability of our Wasm package manager and the usability of other programs that interact with lots of less common ABIs.

The assumption we made is that if a module fits this interface then it was probably written in a higher level language with a library providing the definitions or even an entire wrapper (which could be generated by ⛄️ bindings), so things like passing the correct non-core Wasm types are already taken care of.

The way we handled changes in data types (which are not represented in our interface text format) with identical imports/exports is just to version the interfaces.

Hopefully that gives context into where we're coming from.

To get more context on this: is there a reason this IDL proposal is specific to WASI?  It makes sense in the context of unblocking WASI's split into submodules (which I'm super excited about!), but isn't this IDL more widely applicable?

wanderer commented:
@sbc100 okie I removed method for now. Maybe we can reconsider it once we have the story on dynamic dispatching. 

> To get more context on this: is there a reason this IDL proposal is specific to WASI? It makes sense in the context of unblocking WASI's split into submodules (which I'm super excited about!), but isn't this IDL more widely applicable?

@MarkMcCaskey yes i agree that it could be more widely applicable. I think we will probably eventual merge some efforts with the webild-binding work.

sunfishcode commented:
Thanks @wanderer for your work here! Much of the contents of this WatIDL PR helped shape and contribute to #74 and the witx file format, which has now landed. While witx does currently have a smaller scope than watidl here, it's just a beginning, and we can now work on extending it to better support OCAP in parallel with our other WASI API design work.
